### ISSUE TITLE
[Bug]: Explicit VLLM_CUTLASS NVFP4 MoE selection silently produces empty completions on SM_120

### ISSUE THREAD

**Issue by AubinGil** [vllm-project/vllm#51525]
### Summary

On SM_120 (RTX 5090 Laptop), serving an NVFP4 MoE checkpoint with
`--moe-backend cutlass` (which resolves to `VLLM_CUTLASS`) makes the server
consume the entire `max_tokens` budget and return `content: null` with
`finish_reason: "length"` for certain inputs. There is no exception, no warning,
and the response is HTTP 200. Switching **only** `--moe-backend` to `marlin`
serves the identical requests correctly.

**Severity is bounded, and I want to be upfront about it rather than have you
discover it:** `--moe-backend auto` does *not* select this backend on my machine
(it selects `FLASHINFER_CUTLASS`), so nobody lands here by default. Reaching this
state requires explicitly passing `--moe-backend cutlass`. This is a config
footgun rather than silent corruption in default deployments.

The checkpoint [model card](https://huggingface.co/bg-digitalservices/Gemma-4-26B-A4B-it-NVFP4)
explicitly says that `--moe-backend marlin` is required for correct MoE
computation. I therefore see two possible resolutions: if `VLLM_CUTLASS` is
expected to support this checkpoint layout, this is an output-correctness bug;
if it is incompatible by design, vLLM should reject the explicit selection or
emit a prominent startup warning rather than accept it and silently return
corrupted output.

It still seems worth fixing, for two reasons: the opt-in is a plausible one (see
the selector note at the bottom), and once taken, the failure is invisible — a
short smoke prompt passes, throughput looks normal, and a system prompt fails at
49 tokens. I had a server in this state serving ~2.5 hours of apparently healthy
traffic at up to ~80 tok/s before I noticed, because the server log records
throughput, not content.

### Platform caveat, stated up front

**All of this is on a native-Windows vLLM build.** I know that is not an
officially supported platform, and I cannot rule out that it is Windows-specific.
I could not produce a clean Linux comparison: my WSL install is 0.20.0, and a
stock Linux 0.24.0 wheel hits a quantized tied-embedding gap on this checkpoint
before execution reaches the MoE path.

**If someone has a Linux SM_120 box, confirming or refuting reproducer 1 below
requires one model load plus two requests and would settle this.** I would be
glad to be told this is a Windows build artifact.

I did not modify vLLM's CUTLASS, MoE, quantization, or model-execution code in the
native-Windows environment. (I have a separately patched *Linux* wheel used for
other experiments; it is not involved here.)

### Environment

- vLLM `0.24.0`, `system_fingerprint: vllm-0.24.0-31e2b045`, native Windows
- torch `2.11.0+cu130`, CUDA 13.0
- RTX 5090 Laptop, SM_120
- Model: `bg-digitalservices/Gemma-4-26B-A4B-it-NVFP4`, `--quantization modelopt`
- `--kv-cache-dtype fp8 --max-model-len 16000 --max-num-seqs 1
  --max-num-batched-tokens 4096 --optimization-level 2
  --gpu-memory-utilization 0.93`
- `VLLM_DISABLED_KERNELS=FlashInferFP8ScaledMMLinearKernel,FlashInferCutlassNvFp4LinearKernel`
- `VLLM_USE_FLASHINFER_SAMPLER=0`
- All requests `temperature: 0`

Complete failing-arm launch command:

```powershell
$env:VLLM_DISABLED_KERNELS = "FlashInferFP8ScaledMMLinearKernel,FlashInferCutlassNvFp4LinearKernel"
$env:VLLM_USE_FLASHINFER_SAMPLER = "0"

C:\v\024\Scripts\python.exe -m vllm.entrypoints.openai.api_server `
  --model bg-digitalservices/Gemma-4-26B-A4B-it-NVFP4 `
  --model-impl vllm --host 0.0.0.0 --port 8000 `
  --download-dir D:\huggingface_cache --quantization modelopt `
  --kv-cache-dtype fp8 --moe-backend cutlass --linear-backend cutlass `
  --trust-remote-code --max-model-len 16000 --gpu-memory-utilization 0.93 `
  --max-num-seqs 1 --max-num-batched-tokens 4096 --optimization-level 2 `
  --log-error-stack --enable-auto-tool-choice `
  --tool-call-parser gemma4 --reasoning-parser gemma4
```

Backend selection, failing arm:

```
Using CutlassNvFp4LinearKernel for NVFP4 GEMM
Using 'VLLM_CUTLASS' NvFp4 MoE backend out of potential backends:
  ['FLASHINFER_TRTLLM', 'FLASHINFER_CUTEDSL', 'FLASHINFER_CUTEDSL_BATCHED',
   'FLASHINFER_CUTLASS', 'VLLM_CUTLASS', 'MARLIN', 'EMULATION']
```

Working arm differs only in the second line (`'MARLIN'`). The dense/linear kernel
is `CutlassNvFp4LinearKernel` in **both** arms, so this is the MoE path
specifically, not CUTLASS generally.

### Reproducer 1 — a system-role turn (46 prompt tokens)

```json
{"model": "bg-digitalservices/Gemma-4-26B-A4B-it-NVFP4",
 "messages": [{"role": "system", "content": "You are a terse assistant."},
              {"role": "user", "content": "What is the capital of France? Answer in one short sentence."}],
 "max_tokens": 48, "temperature": 0}
```

Response:

```
"message": { "role": "assistant", "content": null, "refusal": null,
             "annotations": null, "audio": null, "function_call": null,
             "reasoning": null }
finish_reason: "length"   stop_reason: null
usage: prompt_tokens 46, completion_tokens 48, total_tokens 94
```

`reasoning` is null too, so this is not a reasoning parser diverting the text.
Fails 3/3. With `--moe-backend marlin`, the same request returns
`"The capital of France is Paris."`, `finish_reason: stop`, 8 tokens.

It is the **role**, not the text and not the message count. A separate role-isolation matrix used a 49-token prompt:

| variant | prompt tok | result |
|---|---|---|
| `system` role + user | 49 | **empty** |
| `developer` role + user | 49 | **empty** |
| same text moved inside the user turn | 33 | ok |
| `user` / `assistant` / `user` (3 turns) | 49 | ok |

### Reproducer 2 — near-periodic token content, and it is not length

A prompt built from an exact repeating token cycle fails, while unique content of
the same or greater length passes:

| content | prompt tok | result |
|---|---|---|
| unique random words | 46 → 2626 (11 points) | ok at every point |
| repeated sentence ×80 | 666 | **empty** 3/3 |
| repeated sentence ×260 | 2628 | **empty** 3/3 |

So prompt length is not the variable. To separate ordering from token identity, I
held content at 640 tokens and varied the fundamental period. Exact cycles of
period 1, 2, 3, 4, 8, 16, 32, 64 all produced empty output; a deterministic
shuffle of the same tokens passed.

**The shuffle comparison is only exactly controlled for periods 2–16.** There the
shuffled prompt re-tokenized to the identical 640 tokens with the identical
unique-token count, so cycle and shuffle differ *only* in order. At periods 32 and
64 the shuffle re-tokenized differently (652 tok / 52 unique and 654 / 101, versus
640 / 32 and 640 / 64), because reordering words moves word boundaries — those two
pairs are suggestive but confounded and I would not lean on them. Period 1 is
excluded by construction, since shuffling one repeated token is a no-op.

On the five clean pairs, token count, unique-token count and histogram are all
constant and only order changes, so within that range **token order is what
matters**, not token content.

Onset varies with period: period 1 passes to 384 tokens and fails at 512; periods
2, 8, 32 pass at 128 and fail at 256; period 64 passes at 256 and fails at 384.
Phase rotation and ±1 length changes did not fix the 640-token failures.

### What I could not explain

- **No root cause.** I have not traced this into the kernel; I only have
  input/output behaviour and the backend selection lines.
- **A single mutation** in the cycle fixed periods 1 and 64 but not 2–32, so the
  honest claim is "long low-complexity / near-periodic ordering", not that strict
  global periodicity is required.
- Whether it reproduces on another NVFP4 MoE checkpoint, or on SM_100.

### Selector note — why the opt-in is plausible

With both backend flags at `auto`, the selector chose `FLASHINFER_CUTLASS` for
MoE, which then attempted a JIT build that failed on this machine
(`ninja: build stopped: subcommand failed`, `CreateProcess failed: The system
cannot find the file specified`) and aborted `EngineCore`. The natural workaround
for that crash is to pin a CUTLASS backend explicitly — which is exactly how one
arrives at `--moe-backend cutlass` and the silent path above.

That JIT/abort behaviour is the selector asymmetry I described in a comment on
#43507; this issue is only about the *output correctness* of `VLLM_CUTLASS` once
selected, which seemed different enough in scope to warrant its own issue rather
than a third comment on that thread.

### What would help

1. **Fail loudly instead of silently.** Whatever the answer on kernel support,
   vLLM accepted an explicit backend selection and then returned HTTP 200 with
   empty content. If `VLLM_CUTLASS` is known-incompatible with this checkpoint
   layout, refusing the selection or emitting a startup warning would have saved
   this entirely — and nothing in the observed failure points a user back at the
   model card. This is the part I would most like fixed, and it does not depend on
   resolving the question below.
2. Whether `VLLM_CUTLASS` is expected to be a supported selection for NVFP4 MoE on
   SM_120 at all.
3. Confirmation or refutation on **Linux, SM_120** — one model load plus the two
   backend requests above.

I am happy to run further probes on this machine; the scripts and raw JSON results
behind every table above are reproducible on request.

**Comment by taking-lying-flat** [comment]
@AubinGil Confirmed on Linux: this is not a native-Windows artifact.

I reproduced the system+user failure on an RTX 5090 (SM120) with vLLM 0.24.0 and the same `bg-digitalservices/Gemma-4-26B-A4B-it-NVFP4` checkpoint. `VLLM_CUTLASS` returned 48 pad tokens and stopped at the token limit; changing only the MoE backend to Marlin returned `The capital of France is Paris.`

The root cause is an incomplete ModelOpt quantization checkpoint rather than a general CUTLASS kernel failure. Twenty-five experts across 12 layers are missing activation `input_scale` entries; layer 0 is missing `[40, 42, 82, 98]`. These are experts that were apparently never routed during calibration. vLLM allocates the scale buffers with `torch.empty`, partially loads the checkpoint, and then lets the activation-quantized CUTLASS path consume the uninitialized slots without checking completeness. That makes the corruption prompt-dependent because it appears only when routing reaches one of the affected experts. Marlin works because its weight-only path does not consume those activation scales.

This overlaps #45212 and the existing fix in #45320. I posted the Linux evidence and one important validation detail there: checking only for zero/non-finite values is not reliable while the buffers are initialized with `torch.empty`; a NaN load sentinel is needed to detect every missing slot deterministically.

The proposed behavior is fail-fast by default with a clear exception naming the missing expert slots. If maintainers want a lower-priority, explicitly enabled recovery mode, median-filling missing activation scales produced correct output on this reproducer 3/3, but it must emit a prominent warning and cannot guarantee model-wide accuracy. Missing weight scales should remain a hard error.

**Comment by taking-lying-flat** [comment]
@AubinGil Here are two concrete implementation options, including production-code sketches against the current `ModelOptNvFp4FusedMoE` path.

Credit for the existing work:

- #45212 by @rmarnold identified that experts skipped during ModelOpt calibration can be absent from the checkpoint, then traced the resulting `1 / scale` to inf/NaN and pad-token output.
- #45320 by @waynehacking8 contributed the four-parameter validation design, expert-id diagnostics, unconditional weight-scale validation, and the important Marlin/W4A16 activation-scale exemption.
- The Linux SM120 reproduction here confirms the same diagnosis for this Gemma 4 checkpoint. The additional issue is that current `main` uses `torch.empty`, so a missing slot is not guaranteed to remain zero or non-finite.

## Option 1: deterministic fail-fast (preferred)

First import the backend enum:

```python
from vllm.model_executor.layers.fused_moe.oracle.nvfp4 import (
    NvFp4MoeBackend,
    convert_to_nvfp4_moe_kernel_format,
    # ...
)
```

Initialize every per-expert scale that must be loaded from the checkpoint with a NaN sentinel rather than `torch.empty`:

```python
w13_weight_scale_2 = PerTensorScaleParameter(
    data=torch.full(
        (num_experts, w13_num_shards),
        torch.nan,
        dtype=torch.float32,
    ),
    weight_loader=weight_loader,
)

w2_weight_scale_2 = PerTensorScaleParameter(
    data=torch.full((num_experts,), torch.nan, dtype=torch.float32),
    weight_loader=weight_loader,
)

w13_input_scale = PerTensorScaleParameter(
    data=torch.full(
        (global_sf_num_experts, w13_num_shards),
        torch.nan,
        dtype=torch.float32,
    ),
    weight_loader=weight_loader,
)

w2_input_scale = PerTensorScaleParameter(
    data=torch.full(
        (global_sf_num_experts,),
        torch.nan,
        dtype=torch.float32,
    ),
    weight_loader=weight_loader,
)
```

Validate before any backend conversion or reciprocal-scale computation:

```python
def _validate_loaded_expert_scales(self, layer: RoutedExperts) -> None:
    scale_names = ["w13_weight_scale_2", "w2_weight_scale_2"]
    if self.nvfp4_backend != NvFp4MoeBackend.MARLIN:
        scale_names += ["w13_input_scale", "w2_input_scale"]

    for name in scale_names:
        scale = getattr(layer, name)
        invalid_slots = ~torch.isfinite(scale) | (scale <= 0)
        if not invalid_slots.any():
            continue

        invalid_experts = (
            invalid_slots.reshape(scale.shape[0], -1)
            .any(dim=-1)
            .nonzero(as_tuple=True)[0]
            .tolist()
        )
        invalid_indices = invalid_slots.nonzero(as_tuple=False).tolist()
        raise ValueError(
            "Incomplete ModelOpt NVFP4 MoE checkpoint: "
            f"'{name}' has missing or invalid scale values for expert ids "
            f"{invalid_experts} at indices {invalid_indices}. This usually "
            "means quantization calibration did not route tokens to every "
            "expert. Re-export the checkpoint with complete per-expert "
            "scales."
        )
```

Call it at the beginning of `process_weights_after_loading`:

```python
def process_weights_after_loading(self, layer: RoutedExperts) -> None:
    self._validate_loaded_expert_scales(layer)
    # Existing conversion follows.
```

This is preferable as the default: it deterministically catches every missing slot and turns silent HTTP 200 corruption into an actionable startup error. Weight scales remain mandatory on every backend; activation scales are skipped only for Marlin because Marlin does not consume them.

## Option 2: explicit median-repair fallback

If maintainers want a best-effort recovery mode, keep it opt-in and repair only NaN-sentinel activation slots. Compute the median independently for each projection/shard:

```python
def _repair_missing_input_scales(self, layer: RoutedExperts) -> None:
    for name in ("w13_input_scale", "w2_input_scale"):
        scale = getattr(layer, name)
        flat_scale = scale.reshape(scale.shape[0], -1)
        missing = torch.isnan(flat_scale)
        if not missing.any():
            continue

        missing_indices = missing.nonzero(as_tuple=False).tolist()
        with torch.no_grad():
            for shard in range(flat_scale.shape[1]):
                column = flat_scale[:, shard]
                shard_missing = missing[:, shard]
                valid = column[~shard_missing]
                valid = valid[torch.isfinite(valid) & (valid > 0)]
                if valid.numel() == 0:
                    raise ValueError(
                        f"Cannot repair '{name}' shard {shard}: "
                        "no valid loaded scales are available."
                    )
                column[shard_missing] = valid.median()

        logger.warning(
            "Repaired missing ModelOpt NVFP4 activation scales in %s at "
            "indices %s using per-shard medians. Accuracy is not guaranteed.",
            name,
            missing_indices,
        )
```

The call order for an explicitly enabled repair mode would be:

```python
if repair_missing_input_scales:
    self._repair_missing_input_scales(layer)
self._validate_loaded_expert_scales(layer)
```

Important constraints:

- Do not enable repair silently or by default.
- Never impute `w13_weight_scale_2` or `w2_weight_scale_2`; missing weight scales remain a hard error.
- Always run the validator after repair so zero, negative, infinite, or unrepaired values still fail.
- Emit the repaired slot indices and an explicit accuracy warning.

On the full 26B checkpoint, filling only the missing activation-scale slots with per-projection medians restored the CUTLASS reproducer 3/3. That demonstrates a useful fallback, but not model-wide accuracy, so Option 1 should remain the primary behavior.

**Comment by AubinGil** [comment]
@taking-lying-flat Thank you — this is a complete root cause, and the Linux reproduction answers the biggest open question in my report: it is not a native-Windows/WDDM artifact.

The incomplete-checkpoint explanation fits my observations well:

- **Prompt-dependence.** My two triggers (a system-family role turn; long near-periodic content) never made sense as kernel-input properties. As routing patterns that happen to reach one of the 25 uncalibrated experts, they do — including the one result I flagged as unexplained (`p8_break_every128` passing while other break intervals fail), which no longer needs a kernel-level explanation.
- **The Marlin control.** Marlin passing all three reproducers is exactly what a weight-only path that never consumes the missing activation scales would do. The two-arm design pointed at the right boundary; your checkpoint inspection identified what actually sits on it.
- **The model card.** The checkpoint's own "`--moe-backend marlin` required for correct MoE computation" note now reads as the publisher knowing the export was incomplete rather than documenting a kernel limitation.

One nuance worth keeping in the record: this also means my report's title is narrower than the underlying defect. The failure is not specific to `VLLM_CUTLASS` — any backend that consumes per-expert activation scales should be affected by the same holes, and Marlin-class backends should not. My probes only ever varied CUTLASS vs Marlin, so I can't confirm that empirically, but it falls out of your diagnosis.

I agree fail-fast is the right default, and the NaN-sentinel point is important — with `torch.empty` allocation, a zero/non-finite check passing proves nothing. Glad to see 0cc53a5 on #45320 adopt exactly that design (sentinel init, validation naming the parameter and expert ids, Marlin exempted from the activation-scale check).

Happy to verify the PR branch against this checkpoint on my native-Windows SM_120 setup — both that the ValueError fires at load and that it names the expert set you found (layer 0: 40, 42, 82, 98). This issue can close as a duplicate of #45212 / superseded by #45320 whenever maintainers prefer; the reproducers and probe scripts remain available if useful for a regression test.
